src/shared/Microsoft.AzureRepos/AzureReposHostProvider.cs

```diff
@@ -461,6 +493,89 @@ private bool UsePersonalAccessTokens()
             return defaultValue;
         }
 
+        private bool UseServicePrincipal(out ServicePrincipalIdentity sp)
+        {
+            if (!_context.Settings.TryGetSetting(
+                    AzureDevOpsConstants.EnvironmentVariables.ServicePrincipalId,
+                    Constants.GitConfiguration.Credential.SectionName,
+                    AzureDevOpsConstants.GitConfiguration.Credential.ServicePrincipal,
+                    out string spStr) || string.IsNullOrWhiteSpace(spStr))
+            {
+                sp = null;
+                return false;
+            }
+
+            string[] split = spStr.Split(new[] { '/' }, count: 2);
+
+            if (split.Length < 2 || string.IsNullOrWhiteSpace(split[0]))
+            {
+                _context.Streams.Error.WriteLine("error: unable to use configured service principal - missing tenant ID in configuration");
```

Comment:
Nit: if `split.Length` is less than 1, do we know for sure that the tenant ID is missing? Couldn't the client ID be the piece that is missing?

Reply:
True! Added guards and specific messages for missing tenant or client ID.